components/scream/CMakeLists.txt

```diff
@@ -27,6 +27,37 @@ endfunction ()
 enable_testing()
 include(CTest)
 
+# Build kokkos submodule if user did not specify Kokkos_DIR
+if (NOT DEFINED Kokkos_DIR)
+  # Bring in Kokkos
+  set(SCREAM_KOKKOS_CONFIG -D CMAKE_BUILD_TYPE=Release -D KOKKOS_ENABLE_AGGRESSIVE_VECTORIZATION=ON -D KOKKOS_ENABLE_OPENMP=ON -D KOKKOS_ENABLE_PROFILING=OFF -D KOKKOS_ENABLE_DEPRECATED_CODE=OFF -D KOKKOS_ENABLE_EXPLICIT_INSTANTIATION:BOOL=OFF CACHE STRING "How to configure kokkos (generate_makefile args)")
+
+  set(KOKKOS_SRC    ${CMAKE_SOURCE_DIR}/../../externals/kokkos)
+  set(KOKKOS_BINARY ${CMAKE_BINARY_DIR}/kokkos/install)
+
+  file(MAKE_DIRECTORY ${CMAKE_BINARY_DIR}/kokkos)
+  execute_process(
+    COMMAND cmake ${SCREAM_KOKKOS_CONFIG} -DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER} -DCMAKE_INSTALL_PREFIX=${KOKKOS_BINARY} ${KOKKOS_SRC}
+    WORKING_DIRECTORY ${CMAKE_BINARY_DIR}/kokkos
+    ERROR_VARIABLE kokkos_config_ev
+    RESULT_VARIABLE kokkos_config_res)
+  if(NOT kokkos_config_res STREQUAL "0")
+    message(FATAL_ERROR "kokkos gen makefile failed: ${kokkos_config_ev}")
+  endif()
+
+  execute_process(COMMAND make -j8 WORKING_DIRECTORY ${CMAKE_BINARY_DIR}/kokkos ERROR_VARIABLE kokkos_build_ev RESULT_VARIABLE kokkos_build_res)
```

Comment:
I want to think about this a bit. Lines like `make -j8` make me think this mechanism isn't quite right yet. Ideally, Kokkos would (re)build (if needed) as part of the `make` the user issues.

Does Kokkos's CMakeLists.txt tolerate add_subdirectory? If so, I think I'd like to know more reasons _not_ to use it.

Reply:
Yes, it's a bit awkward. The problem is that the line

```
include(Kokkos)
```

Won't work unless Kokkos has been built and installed, but we (SCREAM) need it at configure time.

Reply:
Note, if the user changes their Kokkos configure, that would require them to re-run cmake, which should trigger a Kokkos rebuild. The one case that's not handled very elegantly is if they update their kokkos submodule and then run make without re-running cmake.

Reply:
We can change the `include(Kokkos)` logic, if need be, to behave differently depending on whether _we_ build kokkos or it is used as a tpl. The point that @ambrad makes is good though: I'm not sure if kokkos allows adding it as a subdir. It may assumes that either its main CMakeLists file is the main one or that kokkos is built within trilinos. I have to check. If kokkos allows it, though, it _may_ be the most robust and flexible solution.

Reply:
I could try replacing include(Kokkos) with add_subdirectory if user did not provide Kokkos_Dir.

Reply:
Quickly looking at their main CMakeLists.txt it _seems_ that they do support the use of kokkos as a subdirectory. Worth trying, I think.

Reply:
It looks like Kokkos is not designed to be used via add_subdirectory. The include flags are all wrong. I could potentially hack my way around this problem, but I suspect there would be other problems.

Reply:
I think we can put the 'add_subdirectory' logic in the Kokkos.cmake file. Better handle all the kokkos setup cmake sorcery in a single place, imho.

Reply:
@jgfouca their main CMakeLists.txt does check whether `PROJECT_NAME` is defined, which makes me think that at least they _wanted_ to support this case...

Reply:
I manually added -I${KOKKOS_SRC}/core/src to CXXFLAGS but now I'm hitting issues with KokkosCore_config.h not being available.

Reply:
Dan Ibanez they maybe wanted to support this (add_subdirectory) at one time, but it hasn't been maintained/tested. He says they kinda expect users to build against an installed kokkos.

Reply:
Just FYI, this is what I'm trying now:

```
if (DEFINED Kokkos_DIR)
  include(Kokkos)
  SET (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${KOKKOS_CXXFLAGS_STR}")
else ()
  set(KOKKOS_SRC    ${CMAKE_SOURCE_DIR}/../../externals/kokkos)
  set(KOKKOS_BINARY ${CMAKE_BINARY_DIR}/kokkos/install)

  add_subdirectory(${KOKKOS_SRC} ${KOKKOS_BINARY})

  SET (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -I${KOKKOS_SRC}/core/src")
endif()
```

Unless you guys have some idea where to go from here, I'm going to recommend we go with the PR in its current form for now.